0000-improved-flow-types-for-babel-types.md

```diff
@@ -0,0 +1,164 @@
+|};
+
+....
+
+declare type BabelNode = BabelNodeIdentifier | BabelNodeIfStatement | ...
+```
+
+Nodes can then be refined by the `node.type`
+
+```
+import * as t from '@babel/types';
+import type {StringLiteral, NumericLiteral} from '@babel/types';
+
+function asStringLiteral(ast: ?StringLiteral | NumericLiteral): BabelNodeStringLiteral | null {
+  if (ast == null) {
+      return t.stringLiteral("null");
+  }
+
+  if (ast.type === "StringLiteral") {
+    return ast;
+  }
+
+  return t.stringLiteral(String(ast.value));
+}
+```
+
+[Complete Type Definition](https://gist.github.com/MichaReiser/8d8784b4b9e9d26572f817e44aa6bc62)
+
+# Motivation
+
+The current flow typing defines each node type as a `class` and specifies type refinements (`%checks (node instanceof MemberExpression)`) for the `is*` methods.
+
+```
+const {isIdentifier = require('@babel/types');
+
+if (isIdentifier(node)) {
+  console.log(node.name);
+}
+
+```
+
+The todays system has two shortcomings:
+
+* Exporting the node types as `class`es is semantically incorrect. `@babel/types` does not implement the node types as `class`es nor does it export any value for each type. Declaring the node types as `class`es might misguide users to check for a certain node type using `node instanceof Identifier` or to create a new node by calling the constructor `new Identifier()`.
+* Flow only supports `%checks` refinements on function declarations and that only if the function isn't called as a method. For example, `isIdentifier(node)` is supported but `t.isIdentifier(node)` (method invocation) or `path.isIdentifier()` are not. Having to import each `is*` function is cumbersome and declaring the `is*` methods on `path` isn't possible with what Flow supports today.
+
+
+# Detailed design
+
+The proposal of this RFC is to change today's typing to:
+
+* Use exact types instead of `class`es to make it clear that `@babel/types` does not export values for node types.
+* Use a union type for `BabelNode` rather than a base class. This allows to refine on node types using `node.type === 'Identifier'` or paths `path.node.type === 'Identifier'`.
+
+Example Usage:
+
+```
+import * as t from '@babel/types';
+import type {StringLiteral, NumericLiteral} from '@babel/types';
+
+function asStringLiteral(ast: ?StringLiteral | NumericLiteral): BabelNodeStringLiteral | null {
+  if (ast == null) {
+      return t.stringLiteral("null");
+  }
+
+  if (ast.type === "StringLiteral") {
+    return ast;
+  }
+
+  return t.stringLiteral(String(ast.value));
+}
+```
+
+This changes can be implemented by changing the type definition generation script in `babel/types` for flow only. 
+
+## Regression
+
+Flow does not support a `%check refinement` for types or interfaces. Typing the `is*` methods to refine the `node` type will, therefore, no longer possible.
+
+# Drawbacks
+
+* Braking Change: Applications that use the existing flow types and rely on the `types.is*` methods would have to migrate to `node.type ===` checks.
```

Comment:
This is a big usability regression. While it's easy to replace things like `types.isIdentifier()`, manually inlining functions like `types.isExpression()` or `types.isLoop()` is a mess.

I wouldn't be comfortable with dropping these `%checks` utilities.

Reply:
I agree that this is a regression and I don't know how widely %checks are used. Flow is normally able to infer the %checks itself. Manual %checks predicates are only needed when declaring a function (`declare function ...`) but not when implementing a function.

Reply:
It seems to work with union types :smile: 

https://flow.org/try/#0C4TwDgpgBAqlC8UDeVSQFxQEQywGigFdMBmKAXwG4AoNaANQWVXAky3vygDdMAWCjWoATCAGMANgEMATtG6yoAMwD2KzHAA+UekNGTZ0JYQB2Y4AEsVJqBYDOMABQmVojVG30AlJgBGaiQgpGwBSMQALcQBrOyhnVwgAOjoEeEQcLC8hCyU4+ydVFS8vZGoocqgxazsVQMSJFQBzR0LEwizqcigICTtoJDKKqpMauobm1vaacmogA

Reply:
Yes, `%checks` with `.type` works perfectly. That's what this PR is making use of. The only thing where it gets messy, as you pointed out, is that writing `declare function isLiteral(t) % checks t.type === 'StringLiteral' || t.type === 'NumericLiteral' ...` is somewhat lengthy. I first thought that it was possible in the past to write `declare function isLiteralI(t) %checks t instanceof BabelNodeLiteral` but that wasn't the case since there was no `BabelNodeLiteral` type. 

So the only regressions are that manual `%checks` must be rewritten and that the node types are no longer exported as value which has the implications that:
* Nodes must be imported using `import type`, 
* Node types must be extended using the spread operator instead of `extends` (should be a very rare use case)
* Any other value reference to node types is (correctly) illegal

Reply:
As far as I understand, this RFC doesn't mention that we can still use `%checks` with type. It says that it isn't possible to still use the `is*` methods for type refinements:

> Braking Change: Applications that use the existing flow types and rely on the types.is* methods would have to migrate to node.type === checks.

I see that your PR already keeps the `%checks` in the generated type annotations: could you update this RFC text to match that behavior? You can also write that this is an improvement, since things like `isPattern()` will now help refining.

Reply:
Oh sorry. My fault. I'll update the RFC. It's a relic from an earlier version of the RFC.